ICU-22714 fix zos build, use OpenXL

alexgubanow opened the pull request:
I have fixed the git clone of this repo on zos and moved build from old XLc compiler to new OpenXL compiler.
More details https://unicode-org.atlassian.net/browse/ICU-22714
As of now:
- tests are passing, some issues with files not being UTF-8 on zos, will address that in separate PR
- results of build was tested internally, we have own test suite
- following files had no BOM, need to fix tool that is used to generate those:
icu4c/source/data/misc/currencyNumericCodes.txt
icu4c/source/data/misc/icustd.txt
icu4c/source/data/misc/zoneinfo64.txt

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22714
- [x] Required: The PR title must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [ ] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/unicode-org/icu?pullRequest=3008) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/unicode-org/icu?pullRequest=3008) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/unicode-org/icu?pullRequest=3008) it.</sub>

srl295 commented:
lots of good deletion here!

For the BOMs, would it make sense just to fix them up with a build step?

alexgubanow commented:
OpenXL is basically clang, that allowed me to get rid of many "OS390 only" code and tricks. Actually i have found one bug in OpenXL v2.1, it fails to generate correct name for sidedeck if output path is relative and library name has no suffix, for output library path like:
../../lib/LICU76DA
openxl v2.1 linker will generate ..x name of sidedeck
Fix for that, is already implemented, but no idea when it will be released by IBM, so i made a workaround, added "-Wl,-x$@.x" to linker command, so linker do not guess the name of sidedeck and no bug occur

I have added BOM fix in build time generated index files, also made sure those generated as UTF-8.

But regarding this files:
icu4c/source/data/misc/currencyNumericCodes.txt
icu4c/source/data/misc/icustd.txt
icu4c/source/data/misc/zoneinfo64.txt
i don't know when/who and how updates/generates it. Maybe you guys can guide me there? also i would suggest to do it in separate ticket/PR

alexgubanow commented:
Hello @markusicu @srl295 what is missing for this PR to be merged?

srl295 commented:
@markusicu LGTM ready to merge?

alexgubanow commented:
Hello @markusicu , can we merge this? is anything missing?

markusicu commented on docs/userguide/icu4c/build.md:
```diff
@@ -368,55 +368,54 @@ Some platforms use package management tools to control the installation and unin
 ## How To Build And Install On z/OS (OS/390)
 
-You can install ICU on z/OS or OS/390 (the previous name of z/OS), but IBM tests only the z/OS installation. You install ICU in a z/OS UNIX system services file system such as HFS or zFS. On this platform, it is important that you understand a few details:
+You can install ICU on z/OS or OS/390 (the previous name of z/OS), but IBM tests only the z/OS installation. You install ICU in a z/OS UNIX System Services (z/OS UNIX) file system such as HFS or zFS. On this platform, it is important that you understand a few details:
```
FYI: IBM hasn't testing anything for a long time... :-}

markusicu reviewed (approved):
Sorry for dropping the ball here! Let's get this in.
Happy to see that there are no merge conflicts.
Fingers crossed that the post-merge CI checks will still be green...

markusicu commented:
I had to rebase in order to make a newer CI check work. Now we find out if it still works with the latest checks, even pre-merge...
